web3/_utils/abi.py

```diff
@@ -906,3 +907,80 @@ def build_strict_registry() -> ABIRegistry:
         label="string",
     )
     return registry
+
+
+def named_tree(
+    abi: Sequence[ABIFunctionParams], data: Tuple[Any, ...]
+) -> Dict[str, Any]:
+    """
+    Convert function inputs/outputs or event data tuple to dict with names from ABI.
+    """
+    names = [item["name"] for item in abi]
+    items = [named_subtree(*item) for item in zip(abi, data)]
+
+    return dict(zip(names, items))
+
+
+def named_subtree(
+    # TODO make this a better type
+    abi: Any,
+    data: Tuple[Any, ...],
+) -> Union[Dict[str, Any], Tuple[Any, ...], Any]:
+    abi_type = parse(collapse_if_tuple(dict(abi)))
+
+    if abi_type.is_array:
+        item_type = abi_type.item_type.to_type_str()
+        item_abi = {**abi, "type": item_type, "name": ""}
+        items = [named_subtree(item_abi, item) for item in data]
+        return items
+
+    if isinstance(abi_type, TupleType):
+        names = [item["name"] for item in abi["components"]]
+        items = [named_subtree(*item) for item in zip(abi["components"], data)]
+
+        if len(names) == len(data):
+            return dict(zip(names, items))
+        else:
+            raise MismatchedABI(
+                f"ABI fields {names} has length {len(names)} but received "
+                f"data {data} with length {len(data)}"
+            )
+
+    return data
+
+
+def dict_to_namedtuple(data: Dict[str, Any]) -> Tuple[Any, ...]:
+    def to_tuple(
+        item: Union[Dict[str, Any], List[Any]]
+    ) -> Union[Tuple[Any, ...], List[Any]]:
+        if isinstance(item, dict):
+            return generate_namedtuple_from_dict(**item)
+        return item
+
+    return recursive_map(to_tuple, data)
+
+
+def foldable_namedtuple(fields: Tuple[Any, ...]) -> Tuple[Any, ...]:
+    """
+    Customized namedtuple such that `type(x)(x) == x`.
+    """
+
+    class ABIDecodedNamedTuple(namedtuple("ABIDecodedNamedTuple", fields, rename=True)):  # type: ignore # noqa: E501
+        def __new__(self, args: Any) -> "ABIDecodedNamedTuple":
+            return super().__new__(self, *args)
+
+        def _asdict(self) -> Dict[str, Any]:
+            return dict(super()._asdict())
+
+    return ABIDecodedNamedTuple  # type: ignore
+
+
+def generate_namedtuple_from_dict(**kwargs: Dict[str, Any]) -> Tuple[Any, ...]:
+    """
+    Literal namedtuple constructor such that:
+    `generate_namedtuple_from_dict(x=1, y=2)`
+    returns `generate_namedtuple_from_dict(x=1, y=2)`.
```

Comment:
nit: this comment doesn't make sense to me. Should it be something like: 
```suggestion
    returns `ABIDecodedNamedTuple(x=1, y=2)`.
```

Reply:
removed in refactor